include/lemon/tmalign.hpp

```diff
@@ -346,249 +107,240 @@ struct TMResult {
-//! the notices on the head, the reference information, and this
-//! copyright notice appear in all copies or substantial portions of
-//! the Software. It is provided "as is" without express or implied
-//! warranty.
-//! \param [in] search The frame that is being aligned to native.
-//! \param [in] native The 'native' chain that the search chain is aligned to.
-//! \param [out] rot The aligned version of the search frame.
-//! \param [in] align Should the search frame to aligned to native.
-//! \return A strucure with theTMScore, RMSD after alignment, and number of
-//! aligned residues
-inline TMResult TMscore(const chemfiles::Frame& search,
-                        const chemfiles::Frame& native,
-                        std::vector<chemfiles::Vector3D>& rot,
-                        bool align = false) {
-    std::vector<size_t> a_search;
-    std::vector<size_t> a_native;
+inline TMResult TMscore_helper(const chemfiles::Frame& search,
+                               const chemfiles::Frame& native,
+                               const std::string& s_chain,
+                               const std::string& n_chain) {
 
-    chemfiles::Vector3D t;
-    chemfiles::Matrix3D u = chemfiles::Matrix3D::zero();
+    std::vector<size_t> aligned_search_ids;
+    std::vector<size_t> aligned_native_ids;
 
     // pickup the aligned residues:
-    auto nseqB = count_number_of_atom_names(native, "CA");
-    auto n_ali = find_operlapping_residues(search, native, a_search, a_native);
+    auto n_seq = count_number_of_atom_names(native, "CA");
+
+    // The number of aligned residues
+    auto n_ali = find_operlapping_residues(search, native,
+                                           aligned_search_ids,
+                                           aligned_native_ids,
+                                           s_chain, n_chain);
 
     if (n_ali == 0) {
-        return {0.0, 0.0, 0};
+        return {0.0, 0.0, 0, Affine{{0.0, 0.0, 0.0}, Matrix3D::unit()}};
     }
 
     // parameters:
-    // d0------------->
-    double d0 = 1.24 * std::pow(nseqB - 15, 1.0 / 3.0) - 1.8;
-    if (d0 < 0.5)
-        d0 = 0.5;
-
-    // d0_search ----->
-    double d0_search = d0;
-    if (d0_search > 8)
-        d0_search = 8;
-    if (d0_search < 4.5)
-        d0_search = 4.5;
-
-    // iterative parameters ----->
-    size_t n_it = 20;      // maximum number of iterations
-    size_t n_init_max = 6; // maximum number of L_init
-    size_t L_ini_min = 4;
-    std::vector<size_t> L_ini(n_init_max);
-    if (n_ali < 4)
-        L_ini_min = n_ali;
-
-    size_t n_init;
-    for (n_init = 0; n_init < n_init_max; ++n_init) {
-        if (n_init_max == n_init_max - 1) {
-            L_ini[n_init] = L_ini_min;
-            break;
-        }
+    auto d0 = 1.24 * std::pow(n_seq - 15, 1.0 / 3.0) - 1.8;
+    d0 = std::max(0.5, d0);
+
+    auto d0_search = clamp(d0, 4.5, 8.0);
 
+    // iterative parameters
+    auto max_iter = 20ul;      // maximum number of iterations
```

Comment:
what happen if more than 20 iterations are needed? Is the result of the algorithm still relevant, or should the user be informed?

Reply:
It only means that the algorithm could not find a way to perfectly align all the residues and will exit. The worst result is a bad TMscore, which is reported to the user. Currently, we only get this case in 2 alignments (which I am trying to address before we submit).